Resolve correct link path for tsconfig.extends

zardoy opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

This PR fixes #141112

@mjbvz Sorry for the latency, but its here. I tried, but failed to reuse [existing resolve algorithm](https://github.com/microsoft/TypeScript/blob/febfd442cdba343771f478cf433b0892f213ad2f/src/compiler/commandLineParser.ts#L3005). However I hope it would be possible in future with standalone lib like `tsconfig-paths`.

I think this is almost 1:1 implementation to what we have in TypeScript repo today. Now, if the file doesn't exist, it won't create a link to a non-existing file.
I didin't handle a case where tsconfig might be referenced from the root since its a rare case and I won't believe someone actually does that (however we can check whether path starts with `/` at least).

I made everything up to the provider method async only because of async `exists` (`fs.stat`).

@fregante I remember you compared it to WebStorm. This implementation is much better!

However, I noticed a bug that if you write path to the existing file and then swap it with a non-existing, the provider would be called and it'd return `undefined`, but the old link wouldn't be dropped.

zardoy commented:
I just remembered that I forgot to describe purpose of `resolveNodeModulesPath`. Here is example with library from linked issue:
```txt
- 📁 node_modules
    - 📁 @sindresorhus
        - 📁 tsconfig
            - tsconfig <-- would resolve to here!
- packages/some-package
    - 📁 node_modules
        - 📁 @sindresorhus
            - 📁 another-lib
    - 📁 lib
        - tsconfig.json (extends: @sindresorhus/tsconfig)
```

mjbvz commented on extensions/typescript-language-features/src/utils/fs.ts:
```diff
@@ -14,3 +15,31 @@ export const exists = async (resource: vscode.Uri): Promise<boolean> => {
+	const { fs } = vscode.workspace;
+
+	// clone Uri
+	let currentUri = vscode.Uri.joinPath(currentUriDir, '.');
```
URIs are immutable so you should not need to clone this

mjbvz reviewed (changes requested):
Thanks for taking a look. Added some initial comments. Let me know if you have any questions about them

Can you also please open a new issue to track using node resolution for the tsconfig paths. #131643 was about the referencing the package directory without a `tsconfig.json` at the end, but since this PR adds a lot more functionality we should track that separately

mjbvz commented on extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:
```diff
@@ -16,44 +18,61 @@ function mapChildren<R>(node: jsonc.Node | undefined, f: (x: jsonc.Node) => R):
-				vscode.Uri.file(join(dirname(document.uri.fsPath), extendsNode.value + (extendsNode.value.endsWith('.json') ? '' : '.json')))
-			);
-		}
+		const tsconfigUri = await this.getTsconfigPath(Utils.dirname(document.uri), extendsValue);
```
With this PR, you are potentially performing a lot of file accesses every time the user types in the file. That's not something we can do

The current implementation also lets you create the file if it does not exist when you click the link

zardoy replied:
Okay, in case of relative path we'll always return `.json` path in case if extensionless variant doesn't exist.

But what to do in case if `resolveNodeModulesPath()` returns `undefined`? The previous implementation would just open some path at project's `node_modules`. I don't think user want to create tsconfig at `node_modules`.

I think it makes sense to display error notification to let developer know it. (for example he just forgot to run install command)

zardoy replied:
Also I remember when I was testing it, the link was updating the a HUGE latency. I believe that provider is already called with some latency.

Look, this behavior feels natural: IDE gives a link only when it actually finds a reference, something that points to an actual file.

IMO Creating files via non-existing link is incorrect (I understand that somebody most probably got used to it). There is should be a codefix for appropriate diagnostic instead

zardoy commented:
> Can you also please open a new issue to track using node resolution for the tsconfig paths. #131643 was about the referencing the package directory without a tsconfig.json at the end, but since this PR adds a lot more functionality we should track that separately

I thought #131643 was about incorrect `extends` resolving algorithm in general, but sure I'll do it in a moment.

zardoy commented on extensions/typescript-language-features/src/utils/fs.ts:
```diff
@@ -14,3 +15,31 @@ export const exists = async (resource: vscode.Uri): Promise<boolean> => {
+			}
+		}
+		// reached the root
+		if (posix.relative(currentUri.path, '/') === '') { return; }
```
I can't agree with you. I often use [Monorepo Workspace extension](https://marketplace.visualstudio.com/items?itemName=folke.vscode-monorepo-workspace) to open some package as workspace, I want it resolve correctly in these cases.

Also, git features works in subfolders.

zardoy commented:
- I changed the implementation and added new internal command.

- I didn't add the nls key, I looked into `./rename.ts` and can see it wasn't defined too.

- I tried to test the command, it opens the tsconfig, but doesn't display error message. Breakpoints works from time to time, so I can't figure out things (everything worked perfectly with previous commit).

    - When i try to open nonexisting file it seems that `window.showTextDocument` and `workspace.openTextDocument` complain about in console (`cannot open ... Unable to resolve nonexistent file`)
    - When it returns `undefined` nothing happens

zardoy commented:
> Can you also please open a new issue to track using node resolution for the tsconfig paths

I opened #141112, that has a link reference, feel free to request details if I missed something.

zardoy commented on extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:
```diff
@@ -122,5 +175,15 @@ export function register() {
+		vscode.commands.registerCommand(OPEN_EXTENDS_LINK_COMMAND_ID, async ({ resourceUri, extendsValue, }: OPEN_EXTENDS_LINK_COMMAND_ARGS) => {
+			const tsconfigPath = await getTsconfigPath(Utils.dirname(resourceUri), extendsValue);
+			if (tsconfigPath === undefined) {
+				void vscode.window.showErrorMessage(localize('openTsconfigExtendsModuleFail', "Failed to resolve {0} as module", extendsValue));
```
Added void, as we don't need to wait until user resolve an error.
By adding void we make this line [no-floating-promises](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-floating-promises.md) rule aware. I hope it will enabled in https://github.com/microsoft/vscode/issues/128738. 
Or I should throw it instead?

mjbvz commented on extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:
```diff
@@ -122,5 +175,15 @@ export function register() {
+				vscode.window.showErrorMessage(localize('openTsconfigExtendsModuleFail', "Failed to resolve {0} as module", extendsValue));
+				return;
+			}
+			await vscode.workspace.openTextDocument(tsconfigPath);
```
`openTextDocument` opens the text document but you then need to call `showTextDocument` to show it

mjbvz replied:
Also, you may need to use the `vscode.open` command instead of `openTextDocument` + `showTextDocument`. I believe the command handles the case where the file does not exist and lets users create it if they want (but please test to confirm this)

zardoy replied:
> `openTextDocument` opens the text document but you then need to call `showTextDocument` to show it

I don't really understand the difference between these, though, even my friend [tried](https://github.com/zardoy/vscode-framework/issues/9#issuecomment-951319097) to explaind it. Nothing worked here except the `vscode.open` command.

However the latter just shows error notification instead of modal, but with desired `create file` button!

mjbvz commented on extensions/typescript-language-features/yarn.lock:
```diff
@@ -31,3 +31,8 @@ vscode-nls@^5.0.0:
   resolved "https://registry.yarnpkg.com/vscode-nls/-/vscode-nls-5.0.0.tgz#99f0da0bd9ea7cda44e565a74c54b1f2bc257840"
   integrity sha512-u0Lw+IYlgbEJFF6/qAqG2d1jQmJl0eyAGJHoAJqr2HT4M2BNuQYSEiSE75f52pXHSJm8AlTjnLLbBFPrdz2hpA==
+
+vscode-uri@^3.0.3:
```
Please revert the yarn.lock changes as they cause our CI checks to fail

zardoy replied:
Interesting, whats the point of that workflow?? I reverted the change, but now other actions are failing because of `--frozen-lockfile`

mjbvz commented on extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:
```diff
@@ -110,6 +114,55 @@ class TsconfigLinkProvider implements vscode.DocumentLinkProvider {
+				}
+			}
+			// reached the root
+			if (posix.relative(currentUri.path, '/') === '') { return; }
```
This should stop when we reach the root of the workspace, not the root of the user's file system

zardoy replied:
Is there any reason for doing this? I don't use `getWorkspaceFolder` and `posix.relative(currentUri.path, currentWorkspacePath.uri.path) === ''` on purpose as this should conform Node resolving algorithm, it was a purpose of this PR!

For example what if someone wants to open `packages/some-package` with structure from https://github.com/microsoft/vscode/pull/141062#issuecomment-1017432148? tssever won't stop at root of the workspace, why we should??

Also, you have already started discussion about that that above.

zardoy replied:
Also, WebStorm works this way..

zardoy replied:
I fixed a bug with inifite loop, so now error message displayed correctly. The reason was that `path.relative('C:/', '/')` never gives an empty string.

mjbvz commented on extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:
```diff
@@ -110,6 +114,55 @@ class TsconfigLinkProvider implements vscode.DocumentLinkProvider {
+	const isRelativePath = ['./', '../'].some(str => extendsValue.startsWith(str));
+	if (isRelativePath) {
+		const absolutePath = vscode.Uri.joinPath(baseDirUri, extendsValue);
+		if (await exists(absolutePath)) { return absolutePath; }
```
[nit] Put body of statements on own line

mjbvz commented on extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:
```diff
@@ -110,6 +114,55 @@ class TsconfigLinkProvider implements vscode.DocumentLinkProvider {
+		const absolutePath = vscode.Uri.joinPath(baseDirUri, extendsValue);
+		if (await exists(absolutePath)) { return absolutePath; }
+		// Will suggest to create a .json variant if it doesn't exist yet
+		return absolutePath.with({
```
For a path `./tsconfig.base.json` that doesn't exist on disk, I believe this will end up trying to create a file called `tsconfig.base.json.json`. Instead you should only add `.json` when the path does not already have a file extension

zardoy replied:
You was right! Fixed. I should have better looked at the reference.

mjbvz commented on extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:
```diff
@@ -110,6 +114,55 @@ class TsconfigLinkProvider implements vscode.DocumentLinkProvider {
+	return resolveNodeModulesPath([
+		extendsValue,
+		`${extendsValue}.json`,
+		`${extendsValue}/tsconfig.json`,
```
We should be able to skip some of these checks entirely. For example if the file already ends in `.json`, there is no purpose to checking `${extendsValue}.json` or `${extendsValue}/tsconfig.json`

Failing these early means we can skip going to disk to check if the file exists

zardoy replied:
But in case if we have `module.json/tsconfig.json` in `node_modules` TS can resolving it by referencing `module.json` from extends.
I'm gonna think about it. I'm probably will be there in a day.

zardoy replied:
I made impl a little more complicated, now we skip checks for just `node_modules` and check instead full package path. If the package is not present (with structure from https://github.com/microsoft/vscode/pull/141062#issuecomment-1017432148), it won't check all path candidates within first encountered node_modules and instead 5 checks overall will be performed.
Do you have any other ideas on how it can be optimized or on how to make it cleaner?

zardoy replied:
I wish we had a way to reuse Node's resolving algorithm from somewhere else.

mjbvz commented on extensions/typescript-language-features/src/languageFeatures/tsconfig.ts:
```diff
@@ -110,6 +114,55 @@ class TsconfigLinkProvider implements vscode.DocumentLinkProvider {
+		`${extendsValue}/tsconfig.json`,
+	]);
+
+	async function resolveNodeModulesPath(pathCandidates: string[]): Promise<vscode.Uri | undefined> {
```
[nit] Not a fan of nesting functions definitions like this. Instead move this to the top level of the file and pass in any needed parameters

zardoy replied:
I firstly resolved it because I thought you was talking about moving it above the `return ...` statement. But whats the point of having it defined on top level, since it's being used only in one place? I'm pretty sure this approach reduces readability of your codebase.

TLDR By keeping the definition scoped it is much easier to understand where it's being used. WDYT?

zardoy replied:
Didn't see a reply so I just moved it like you said. I don't mind of almost any codestyle, sorry for arguing. Btw what nit does mean?

mjbvz reviewed (changes requested):
A few more changes but getting closer!

zardoy commented:
Please review all opened comments when you have time for it. I'll be waiting for reply!

mjbvz commented:
Thanks! Will be in the next insiders build for testing
